src/OpenTK.Windowing.Desktop/NativeWindow.cs

```diff
@@ -950,31 +946,14 @@ private unsafe void ProcessInputEvents()
             GLFW.GetCursorPos(WindowPtr, out var x, out var y);
             _mouseState.Position = new Vector2((float)x, (float)y);
 
-            for (var i = 0; i < JoystickStates.Length; i++)
+            for (var i = 0; i < _joystickStates.Length; i++)
             {
-                var joy = JoystickStates[i];
-                if (joy == default)
+                if (_joystickStates[i] == null)
                 {
                     continue;
                 }
 
-                var h = GLFW.GetJoystickHatsRaw(joy.Id, out var count);
-                var hats = new Hat[count];
-                for (var j = 0; j < count; j++)
-                {
-                    hats[j] = (Hat)h[j];
-                }
-
-                var axes = GLFW.GetJoystickAxes(joy.Id);
-
-                var b = GLFW.GetJoystickButtonsRaw(joy.Id, out count);
-                var buttons = new bool[count];
-                for (var j = 0; j < buttons.Length; j++)
-                {
-                    buttons[j] = b[j] == JoystickInputAction.Press;
-                }
-
-                JoystickStates[i] = new JoystickState(hats, axes, buttons, joy.Id, joy.Name);
+                ((JoystickState)_joystickStates[i]).Update();
```

Comment:
Not ideal to be casting, but it's survivable as internal code.

Reply:
Less ideal to put Update() on the interface that represents an "immutable" object.

Reply:
Unless we had `interface IJoystickState : IUpdatable`
Then use IUpdatable on the other input state classes too.

Reply:
This could be done in this function instead using the set state functions.
Or we could also just store these as the internal type.
We could still have `IReadOnlyCollection<IJoystickState>` be the type of the public api as `IReadOnlyCollection<T>` is covariant.